subprojects/gradle/runner/service/src/main/kotlin/com/avito/runner/service/worker/device/adb/AdbDevice.kt

```diff
@@ -163,22 +163,21 @@ data class AdbDevice(
             delaySeconds = 5,
             block = { attempt ->
                 logger.debug("Attempt $attempt: getting boot_completed param")
-                val devicesBooted: Boolean = loadProperty(
+                val bootCompleted: Boolean = loadProperty(
                     key = "sys.boot_completed",
                     cast = { output -> output == "1" }
                 )
-                logger.debug("Attempt $attempt: boot_completed param is $devicesBooted")
+                logger.debug("Attempt $attempt: boot_completed param is $bootCompleted")
 
-                if (!devicesBooted) {
-                    throw RuntimeException(
-                        "Device: $id hasn't booted yet"
-                    )
+                if (!bootCompleted) {
+                    // TODO it's hard to throw exception on each retry
```

Comment:
What's the problem?

Reply:
We create the exception and it's a hard operation because you need to compute stacktrace
I consider we could use lighter way to make decisions

Reply:
> it's a hard operation because you need to compute stacktrace

How did measure that this is a problem?

Anyway, I agree that it's no good to use exceptions for conditionals.